src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -313,56 +320,197 @@ long heapAllocations()
+                // the previewed key.
+                Object keys = keySets.peek();
+                if (keys == null || peek(keys).compareTo(nextKey) >= 0)
+                    break;
+
+                // Now that we know we're skipping past the previewed key, so perform the skip on this keySet.
+                keySets.poll();
+                if (keys instanceof SortedSet)
                 {
-                    mergedKeys.add(first);
-
-                    minimumKey = minimumKey == null ? first : first.compareTo(minimumKey) < 0 ? first : minimumKey;
-                    maximumKey = maximumKey == null ? first : first.compareTo(maximumKey) > 0 ? first : maximumKey;
+                    var newKeys = ((SortedSet<PrimaryKey>) keys).tailSet(nextKey);
+                    if (!newKeys.isEmpty())
+                        keySets.offer(newKeys);
+                }
+                else if (keys instanceof SortedSetRangeIterator)
+                {
+                    var iterator = (SortedSetRangeIterator) keys;
+                    iterator.skipTo(nextKey);
+                    if (iterator.hasNext())
+                        keySets.offer(iterator);
                 }
+            }
+        }
 
-                return;
+        @Override
+        protected PrimaryKey computeNext()
+        {
+            Object keys = keySets.poll();
+            if (keys == null)
+                return endOfData();
+
+            PrimaryKey result = peek(keys);
+            assert result != null;
+
+            SortedSetRangeIterator iterator = dropFirst(keys);
+            if (iterator != null)
+                keySets.offer(iterator);
+
+            return result;
+        }
+
+        @Override
+        public void close() throws IOException
+        {
+        }
+
+        /**
+         * The purpose of this method is to avoid unnecessary allocation of iterators for singleton sets.
+         * If the keys object contains a single key, null is returned.
+         * If the keys object contains more than one key, the first key is dropped and the iterator to the
+         * remaining keys is returned.
+         */
+        private static @Nullable SortedSetRangeIterator dropFirst(Object keys)
+        {
+            if (keys instanceof PrimaryKey)
+                return null;
+
+            SortedSetRangeIterator iterator = (keys instanceof SortedSetRangeIterator)
+                ? (SortedSetRangeIterator) keys
+                : new SortedSetRangeIterator((SortedSet<PrimaryKey>) keys);
+
+            assert iterator.hasNext();
+            iterator.next();
+            return iterator.hasNext() ? iterator : null;
+        }
+
+        static PrimaryKey peek(Object keys)
+        {
+            if (keys instanceof PrimaryKey)
+                return (PrimaryKey) keys;
+            if (keys instanceof SortedSet)
+                return (PrimaryKey) ((SortedSet<?>) keys).first();
```

Comment:
An interesting consequence of deferring the wrap of the `SortedSet` by the `SortedSetRangeIterator` is that the `first()` PrimaryKey could change underneath us, which could result in an out of order PQ and as a consequence, an out of order `RangeIterator`, which could give us correctness problems. That is because we have a direct reference to the `SortedSet` and not a reference to an iterator based on the set. As such, the `SortedSet`'s first PK could change, which isn't true of the iterator. (Though an iterator could have elements added, but it won't have the `head` changed underneath us.)

We might need to wrap all multi-valued SortedSets to the SortedSetRangeIterator. What is your perspective?

Reply:
This is a great catch. Yes, this would be a very subtle issue and I can imagine how it could cause wasting a week of debugging for someone. Surely I need to fix it.

Reply:
Fixed